### PR TITLE
[NPU] Add mxfp4-w4a4 MOE Quantization Support for NPU

### PR DESCRIPTION
**Pull request by LinyuanLi0046**
## Motivation

SGLang already supports ModelSlim W4A4_MXFP4 quantization for dense linear layers on Ascend NPU, but the corresponding MoE path is still missing. As a result, MoE models whose expert weights are exported with the W4A4_MXFP4 scheme cannot be loaded and executed through the ModelSlim quantization backend.
The Ascend MoE implementation has also been refactored into the modular AscendRunner architecture. Therefore, MXFP4 MoE support needs to integrate with the new per-weight-group quantization methods and runner flow instead of restoring the legacy fused MoE implementation.

## Modifications

- Add a ModelSlim W4A4_MXFP4 MoE scheme and register it for MoE quantization detection.
- Create w13 and w2 expert weights independently, following the existing ModelSlim MXFP4 checkpoint format:FP4 values stored in float8_e4m3fn containers.
- UE8M0 block scales stored as uint8 with a block size of 32.

- Add an Ascend MXFP4 MoE quantization method compatible with the refactored AscendRunner and GroupedMatmul interfaces.
- Repack offline FP8-container weights into float4_e2m1fn_x2, convert them to the required NPU layout, and reshape block scales for grouped matmul.
- Dynamically quantize BF16/FP16 activations to MXFP4 before both expert GMM operations.
- Keep the dispatcher output in BF16, since the refactored Ascend dispatcher currently supports BF16 and INT8 outputs, and perform MXFP4 quantization inside the MoE method.
- Treat MoE weight offsets as optional so offset-free schemes such as MXFP4 can use the common ModelSlim MoE runner.
- Reuse the existing routing, activation, and finalization flow provided by the refactored Ascend MoE backend.
- Leave the existing W4A4_MXFP4 linear implementation unchanged.

## Accuracy Tests

Tested with Qwen3-235B-A22B mxfp4 w4a4 weights.
```
 metrics = run_eval(args)
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [11:09<00:00,  1.97it/s]
Accuracy: 0.938
Invalid: 0.000
Latency: 669.312 s
Output throughput: 263.397 token/s
metrics={'accuracy': np.float64(0.9378316906747536), 'invalid': np.float64(0.0), 'latency': 669.312145670061, 'output_throughput': 263.39728203125276}
metrics['accuracy']=np.float64(0.9378316906747536)
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #31941258270](https://github.com/sgl-project/sglang/actions/runs/31941258270)
Latest PR Test (Extra): :x: [Run #31941258179](https://github.com/sgl-project/sglang/actions/runs/31941258179)

### CODE REVIEW

File: python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py [hunk 1]
```diff
@@ -141,6 +163,95 @@ def _get_bias_args(
...
+        super().__init__(quant_config=None)
+        self.matmul = GroupedMatmul()
+
+    def process_weights_after_loading(
+        self, layer: torch.nn.Module, weight_prefix: str
+    ) -> None:
+        self._validate_weight_prefix(layer, weight_prefix)
+
+        weight = getattr(layer, f"{weight_prefix}_weight")
+        weight.data = npu_format_cast(weight.data).transpose(-1, -2)
```
Comment (OrangeRedeng) [inline python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py:296]: The same comment as in w4a8 case, maybe we should use .continues() here

File: python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py [hunk 2]
```diff
@@ -141,6 +163,95 @@ def _get_bias_args(
...
 
+# ---------------------------------------------------------------------------
+#  NPUW4A4MXFP4MoEMethod
+# ---------------------------------------------------------------------------
+class NPUW4A4MXFP4MoEMethod(_NPUMoEMethodBase):
+    """ModelSlim W4A4 MXFP4 MoE with single-level FP4 weights and activations."""
+
+    def __init__(self):
+        super().__init__(quant_config=None)
+        self.matmul = GroupedMatmul()
```
Comment (OrangeRedeng) [inline python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py:281]: The same comment as in w4a8, could we use GroupedMatmulSwigluQuant here to improve performance?

File: python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py [hunk 3]
```diff
@@ -141,6 +163,95 @@ def _get_bias_args(
...
+                2,
+            )
+            .transpose(1, 2)
+        )
+        weight_scale.data = scale
+
+        # The refactored Ascend dispatchers currently support BF16 and INT8.
+        # Keep dispatch in BF16 and quantize immediately before each GMM.
+        if weight_prefix == "w13":
+            self._set_dispatcher_output_dtype(layer, "bf16")
```
Comment (OrangeRedeng) [inline python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py:310]: Could we implement mxfp4 quant for init_moe_routing? It seems that kernel now support it https://www.hiascend.com/document/detail/zh/CANNCommunityEdition/latest/API/aolapi/context/ops-transformer/aclnnMoeInitRoutingV3.md

File: python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py [hunk 4]
```diff
@@ -141,6 +163,95 @@ def _get_bias_args(
...
+
+        if pertoken_scale is None:
+            hidden_states, pertoken_scale = torch.ops.npu.npu_dynamic_mx_quant(
+                hidden_states,
+                axis=1,
+                round_mode="rint",
+                dst_type=fp4_dtype,
+                block_size=MXFP4_BLOCK_SIZE,
+                scale_alg=None,
+            )
```
Comment (OrangeRedeng) [inline python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py]: The same comment as for w4a8, please could you move this code to HiddenStatesDynamicQuant (just like it's done for mxfp8?)

**Review by OrangeRedeng — COMMENTED** [review]
Hi! Thank you for your contribution! I left the same as for w4a8 comments, which primarily focus on improving performance and using unified classes. And if it's not too much trouble, could you provide the results of the accuracy checks?

File: python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py (hunk 1 above)
Comment (OrangeRedeng) [reply python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py:296]: But I'm not sure, since I had some problems with w4a8 mxfp with continues() tensors, I think it's worth leaving everything as it is for now

**Comment by ping1jing2** [comment]
please remember post your test result here

File: python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py (hunk 4 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py]: fixed

**Comment by LinyuanLi0046** [comment]
> please remember post your test result here

added

File: python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py (hunk 3 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py:310]: Tests show that based on performance data and profiling operator latency, this approach provides no obvious performance gains. Therefore, to keep the changes concise, I won't add it. If you later find a scenario where it does improve performance, you can submit a separate PR

File: python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py (hunk 2 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py:281]: Tests show that based on performance data and profiling operator latency, this approach provides no obvious performance gains. Therefore, to keep the changes concise, I won't add it. If you later find a scenario where it does improve performance, you can submit a separate PR

File: python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py (hunk 1 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/quantization/moe_methods.py:296]: continues() on FP4 NZ tensors will lead to some errors

**Review by OrangeRedeng — APPROVED** [review]

**Comment by OrangeRedeng** [comment]
@LinyuanLi0046 Hi, could you please fix merge conflicts

**Comment by LinyuanLi0046** [comment]
> @LinyuanLi0046 Hi, could you please fix merge conflicts

done

**Comment by OrangeRedeng** [comment]
Ci analysis: 
- Extra (x2): gate-only failures (`run-ci-extra` label is missing); no Extra tests ran.
- Arm64: `sgl_kernel` export/unit-test failure. Build and ARM sanity check passed. Unrelated to MXFP4 changes.
- Xeon: `base-b-test-cpu` unit-test failure. Build and AMX checks passed. Unrelated to MXFP4 changes.
- XPU: stage A passed; stage B tests failed. `finish` is only a cascading failure. Unrelated to MXFP4 changes.
- ROCm 7.2: ROCm-specific test failure, outside the NPU/Ascend MXFP4 path.
- NPU: only `multimodal-gen-test-1-npu-a3` and `base-c-test-perf-8-npu-a3` failed. Other base/accuracy tests and perf 2/4/16 passed. These are generic multimodal/perf failures and do not exercise the new MXFP4 MoE path; `pr-test-npu-finish` is only cascading.

All current failures appear unrelated to the MXFP4 MoE changes.

**Review by ping1jing2 — APPROVED** [review]
